text/0000-confidential-token-ids.md

```diff
@@ -0,0 +1,270 @@
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+### Masked Token ID Field in the TXO
+
+A `TokenType` is added which specifies a `u32 token_id`. The `token_id` for each TXO represents the Token Type of that TXO, with the first `TokenType` being `MOB = token_id(0)`. New TokenTypes increase the `token_id` monotonically. For example, the `TokenType` struct may be something like,
+
+```
+struct TokenType {
+
+    \\\ Token Name, such as MOB
+    string token_name,
+    
+    \\\ Token ID to be represented confidentially in each TXO
+    u32 token_id,
+    
+    \\\ Curve base point, unique for each TokenType, from which the TXO's commitment is 
+    \\\ generated with the value of the `amount` in the TXO
+    RistrettoPoint H_i
+}
+```
+
+The `masked_token_id` field is added to the [`Amount`](https://github.com/mobilecoinfoundation/mobilecoin/blob/master/transaction/core/src/amount/mod.rs#L28) in each TXO, and functions similar to `masked_value`. The protocol representation of the TXO uses Protobuf, which enables backward compatability in this way: if the field is not populated (Protobuf bytes field of length 0), the default `token_id` is 0. Otherwise, the field is a Protobuf bytes field of length 4. Please see the [Upgrade Plan](#upgrade-plan) for more discussion on rolling out this change with consideration for backward compatibility.
+
+- To compute the 4 byte `masked_token_id` from the `token_id`, we hash the TXO shared secret, with a prefix, XORed with the `token_id`, and take the little endian representation of those bytes.
+
+    ```
+    masked_token_id = (token_id ^ Blake2B(token_id_tag | shared_secret)).to_le()
+    ```
+
+- To recover the `token_id` from the `masked_token_id`, we reverse the process (if the bytes field is length = 4, otherwise `token_id = 0`).
+
+    ```
+    unmasked_token_id = masked_token_id ^ Blake2b(token_id_tag | shared_secret)
+    ```
+
+### Proof of Opening: Proof of Knowledge of Uniform Token ID in the Transaction
+
+We would like to be able to prove that all transaction inputs and outputs are using the same `token_ids`. We do this using a non-interactive, zero knowledge Proof of Opening that the sender has the capability to unwrap the sum of output commitments in the transaction, in conjunction with the homomorphic encryption property of [Pedersen Commitments](https://www.cs.cornell.edu/courses/cs754/2001fa/129.PDF), namely that addition on the commitments preserves the additive relationship of the pre-committed values, as long as they are using the same group generator.
+
+- To prove each TXO's commitment to the `amount` value, we use a Pedersen commitment, consisting of a group generator (`H_i`) raised to the value of the `amount`, with a blinding factor to prevent brute-forcing the full range of values (`2^64`). We compute the blinding factor as the blinding group generator raised to a hash of the shared secret for the TXO.
```

Comment:
Honestly I am sort of confused by this 'Proof of Opening' approach. You need a range proof on all output amount commitments, so I'm pretty sure it's sufficient to say: 1) all output amount commitments are range-proofed on `H_i`, 2) sum(inputs) == sum(outputs). The range proofs on outputs will show that `sum(outputs) = _ G + _ H_i` (for some unimportant values `_`), which implies `sum(inputs) == _ G + _ H_i` (all inputs were previously range proofed, which means it isn't possible for a sum of inputs to overflow - the only way you could have `sum(inputs) == _ G + _ H_i + _ H_n`).

Reply:
Hmm, I think you are right, I think the thing being proved by the proof of opening is actually already implied by the range proofs being over `H_i`. I'm trying to figure out if there is some subtle loss in security parameter around that, but I don't see that there is.

Reply:
Good observation! I think you're right that the proof-of-opening is redundant.

Part of me still likes the proof-of-opening, because it directly asserts the transaction requirement that each input value is of the form _G + _ H_i. In my mind, the range proofs exist to prevent numerical overflow. Maybe it's fine if they are also used to assert that all inputs and outputs are of a single token type, but I worry that that kind of implicit claim is the kind of thing that tends to get overlooked if, say, we change the way we implement range proofs, or if we ever decide to use "vector Pedersen commitments" with mixed types (i.e. _G + _ H_i + _ H_k).

I don't feel very strongly about this; just thinking out loud

Reply:
Updated to include @UkoeHB's description in the *Pederesen Generators* section! (It may be too verbose now - very happy to accept edits!)